### DISCUSSION TITLE
Nuget error NU1201 in a WPF project based on .Net Framework 4.8.1

### DISCUSSION THREAD

**Discussion by aries-zhang** [NuGet/Home#12206] [Q&A] [answered]
I have a WPF project built on .Net Framework 4.8 and packaged using [Windows Application Packaging Project][1]. When I tried to upgrade it to .Net Framework 4.8.1 for ARM64 support, I am getting [NU1201][2] error during Nuget restoration.

>NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362). Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)

I have tried to trim everything down to a minimum reproducible repo [here](https://github.com/aries-zhang/WpfApp7): 

There are 2 [tags][3]:

* `netframework48` and 
* `netframework481`

Their difference is switching over to ".Net Framework 4.8.1" in the project properties view:

[![enter image description here][4]][4]

On `netframework48`, run `nuget restore .\WpfApp7.sln` successfully restores the packages, while on `netframework481`, the same command fails with the following error message.

```
> nuget restore .\WpfApp7.sln
MSBuild auto-detection: using msbuild version '17.3.1.41501' from 'd:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Current\bin'.
Assets file has not changed. Skipping assets file writing. Path: D:\Work\temp\WpfApp7\WpfApp7\obj\project.assets.json
Restored D:\Work\temp\WpfApp7\WpfApp7\WpfApp7.csproj (in 69 ms).
Restoring packages for D:\Work\temp\WpfApp7\WpfApp7.Package\WpfApp7.Package.wapproj...
NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362). Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)
NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362) / win10-arm. Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)
NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362) / win10-arm-aot. Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)
NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362) / win10-arm64-aot. Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)
NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362) / win10-x64. Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)
NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362) / win10-x64-aot. Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)
NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362) / win10-x86. Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)
NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362) / win10-x86-aot. Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)
Assets file has not changed. Skipping assets file writing. Path: D:\Work\temp\WpfApp7\WpfApp7.Package\obj\project.assets.json
Failed to restore D:\Work\temp\WpfApp7\WpfApp7.Package\WpfApp7.Package.wapproj (in 500 ms).

Errors in D:\Work\temp\WpfApp7\WpfApp7.Package\WpfApp7.Package.wapproj
    NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362). Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)
    NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362) / win10-arm. Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)
    NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362) / win10-arm-aot. Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)
    NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362) / win10-arm64-aot. Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)
    NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362) / win10-x64. Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)
    NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362) / win10-x64-aot. Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)
    NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362) / win10-x86. Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)
    NU1201: Project WpfApp7 is not compatible with uap10.0.18362 (UAP,Version=v10.0.18362) / win10-x86-aot. Project WpfApp7 supports: net481 (.NETFramework,Version=v4.8.1)

NuGet Config files used:
    C:\Users\arieszhang\AppData\Roaming\NuGet\NuGet.Config
    C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.FallbackLocation.config
    C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.Offline.config

Feeds used:
    C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\
    https://api.nuget.org/v3/index.json
1 of 2 projects are up-to-date for restore.
```

Searched around but still did not figure out what this error exactly means or how to fix this issue? In my project there is neither a project dependency nor a dependency from NetStandard to NetCoreApp as the [definition][2] of the error says.

I have tried every UWP target in the WpfApp7.Package properties, but the error remains regardless which min version selected, so I'm assuming the UWP target version is unlikely the cause. 

[![enter image description here][5]][5]

Same question is also posted here: https://stackoverflow.com/questions/74258798/failing-to-upgrade-a-wpf-app-from-net-framework-4-8-to-net-framework-4-8-1-due

  [1]: https://learn.microsoft.com/en-us/windows/msix/desktop/desktop-to-uwp-packaging-dot-net
  [2]: https://learn.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1201
  [3]: https://github.com/aries-zhang/WpfApp7/tags
  [4]: https://i.stack.imgur.com/48QDB.png
  [5]: https://i.stack.imgur.com/v1bzY.png

**Comment by zivkan** [comment, marked as answer]
It's a consequence of NuGet not providing "packaging projects" like wapproj a way to say "allow a project reference to any project". Behind the scenes it's implemented by wapproj pretending to be a uap10.0.18362 project, with "asset target fallback" to .NET Framework 4.8(.0). Which means, that the wapproj can use ProjectReference to .NET projects up to 4.8, but not 4.8.1 (because from a .NET project to .NET project reference, it's incompatible).

In other words, packaging projects are implemented in a way to "exploit" NuGet features designed for .NET projects, so when something goes wrong, the error message doesn't make a lot of sense since it's being used in a context very different to the scenario the NuGet feature was designed for.

Therefore, a possible workaround is to set `<AssetTargetFallback>net481</AssetTargetFallback>` in a `PropertyGroup` in the wapproj, but MSBuild evaluates top-down, so you might need to set this after all other `<import` directives, if the wapproj team didn't allow for customers to more easily override the value.

Ultimately for this to work "out of the box", a bug needs to be filed and assigned to the wapproj team, so they can update their msbuild props/targets accordingly. Since wapproj is not open source on github, I recommend you report a problem in VS, or otherwise file feedback on [Visual Studio's Developer Community](https://developercommunity.visualstudio.com/home).

**Comment by aries-zhang** [comment]
Thank you @zivkan for this much detailed answer. It explains everything crystal clear👍.
I have followed your workaround and it works. See code [here](https://github.com/aries-zhang/WpfApp7/tree/fix/nuget-net481).

Also a feedback is raised here: https://developercommunity.visualstudio.com/t/Nuget-error-NU1201-in-a-WPF-project-base/10199241. Hope this issue can be fixed soon.
